<h1> 달러, 나스닥 데이터 분석 테스트 </h1>

- 관련 유저스토리 : 트레이더는 수익률 3%를 위해 달러 N일 이상 상승 시 나스닥 상승 확률을 알기 원한다.
- 데이터: 달러, 나스닥 데이터
  - <b style="color:red">이슈: 데이터 종류 / 기간 / 단위</b>
    - <b>달러: DXY / 2023 / Daily</b>
    - <b>나스닥: IXIC / 2023 / Daily</b>

<h2>데이터 전처리</h2>

- 내용
  - 날짜로 인덱스 변경
  - 날짜(인덱스) 기준으로 오름차순 정렬
  - 상승 여부 컬럼(상승) 추가
  - 변동 비율 컬럼(비율) 추가

In [1]:
import pandas as pd

df_dxy = pd.read_csv("./data/DXY-2023.csv")
df_ixic = pd.read_csv("./data/IXIC-2023.csv")

df_dxy = df_dxy.set_index('날짜', drop=False)
df_dxy = df_dxy.sort_index()
df_dxy["상승"] = df_dxy["변동 %"].map(lambda x: float(x[:-1])>0)
df_dxy["하락"] = df_dxy["변동 %"].map(lambda x: float(x[:-1])<0)
df_dxy["비율"] = df_dxy["변동 %"].map(lambda x: float(x[:-1]))
# df_dxy

df_ixic = df_ixic.set_index('날짜', drop=False)
df_ixic = df_ixic.sort_index()
df_ixic["상승"] = df_ixic["변동 %"].map(lambda x: float(x[:-1])>0)
df_ixic["하락"] = df_ixic["변동 %"].map(lambda x: float(x[:-1])<0)
df_ixic["비율"] = df_ixic["변동 %"].map(lambda x: float(x[:-1]))
# df_ixic

In [2]:
df_dxy

날짜      종가      시가      고가      저가  거래량    변동 %  \
날짜                                                                        
2023- 01- 01  2023- 01- 01  103.49  103.49  103.49  103.49  NaN  -0.03%   
2023- 01- 03  2023- 01- 03  104.52  103.66  104.86  103.47  NaN   0.99%   
2023- 01- 04  2023- 01- 04  104.25  104.58  104.60  103.83  NaN  -0.26%   
2023- 01- 05  2023- 01- 05  105.04  104.07  105.27  103.99  NaN   0.76%   
2023- 01- 06  2023- 01- 06  103.88  105.05  105.63  103.87  NaN  -1.11%   
...                    ...     ...     ...     ...     ...  ...     ...   
2023- 12- 18  2023- 12- 18  102.56  102.59  102.63  102.38  NaN  -0.06%   
2023- 12- 19  2023- 12- 19  102.17  102.47  102.63  102.07  NaN  -0.38%   
2023- 12- 20  2023- 12- 20  102.41  102.19  102.54  102.16  NaN   0.24%   
2023- 12- 21  2023- 12- 21  101.84  102.39  102.45  101.74  NaN  -0.55%   
2023- 12- 22  2023- 12- 22  101.70  101.76  101.89  101.43  NaN  -0.14%   

                 상승     하락    비율  
날짜                                
2023- 01- 01  False   True -0.03  
2023- 01- 03   True  False  0.99  
2023- 01- 04  False   True -0.26  
2023- 01- 05   True  False  0.76  
2023- 01- 06  False   True -1.11  
...             ...    ...   ...  
2023- 12- 18  False   True -0.06  
2023- 12- 19  False   True -0.38  
2023- 12- 20   True  False  0.24  
2023- 12- 21  False   True -0.55  
2023- 12- 22  False   True -0.14  

[256 rows x 10 columns]

In [3]:
df_ixic

날짜         종가         시가         고가         저가  \
날짜                                                                       
2023- 01- 03  2023- 01- 03  10,386.99  10,562.06  10,613.06  10,309.16   
2023- 01- 04  2023- 01- 04  10,458.76  10,467.82  10,515.22  10,337.64   
2023- 01- 05  2023- 01- 05  10,305.24  10,390.31  10,393.22  10,295.25   
2023- 01- 06  2023- 01- 06  10,569.29  10,363.96  10,604.14  10,265.04   
2023- 01- 09  2023- 01- 09  10,635.65  10,662.11  10,807.26  10,619.12   
...                    ...        ...        ...        ...        ...   
2023- 12- 18  2023- 12- 18  14,904.81  14,814.02  14,938.04  14,811.82   
2023- 12- 19  2023- 12- 19  15,003.22  14,923.02  15,003.80  14,921.60   
2023- 12- 20  2023- 12- 20  14,777.94  14,973.36  15,069.29  14,776.29   
2023- 12- 21  2023- 12- 21  14,963.87  14,923.14  14,970.81  14,837.47   
2023- 12- 22  2023- 12- 22  14,992.97  15,006.18  15,047.19  14,927.12   

                  거래량    변동 %     상승     하락    비율  
날짜                                                 
2023- 01- 03  983.36M  -0.76%  False   True -0.76  
2023- 01- 04    1.04B   0.69%   True  False  0.69  
2023- 01- 05  957.48M  -1.47%  False   True -1.47  
2023- 01- 06  970.67M   2.56%   True  False  2.56  
2023- 01- 09    1.04B   0.63%   True  False  0.63  
...               ...     ...    ...    ...   ...  
2023- 12- 18    1.15B   0.61%   True  False  0.61  
2023- 12- 19    1.13B   0.66%   True  False  0.66  
2023- 12- 20    1.20B  -1.50%  False   True -1.50  
2023- 12- 21  945.08M   1.26%   True  False  1.26  
2023- 12- 22  880.44M   0.19%   True  False  0.19  

[246 rows x 10 columns]

- <b>이슈: 달러 휴일 데이터 처리</b>
  - 사용 O
- <b>이슈: 보합 데이터 처리</b>
  - 달러: 보합 조건 X
  - 나스닥: 하락으로 분류

<h2>데이터 분석</h2>

<h3>분석 1: 달러 -> 다음날 나스닥</h3>

In [7]:
# 조건 : DXY, IXIC / 2023(~12.22) / Daily
# !리팩터링 필요 (Dataframe 활용)

dict_dxy = df_dxy.to_dict(orient="records")
dict_dxy_max_index = len(dict_dxy) - 1

def check_dxy_cond(case: str, threshold: int, data: float) -> bool:
    if case == "상승":
        return data >= threshold
    else: # 하락
        return data <= (-1) * threshold

dxy_cases = ["상승", "하락"]
dxy_thresholds = [0,1,2,3,4,5]
ixic_thresholds = [0,1,2,3,4,5]

for dxy_case in dxy_cases:
    for dxy_threshold in dxy_thresholds:
        is_printed = False
        for ixic_threshold in ixic_thresholds:
            
            total = 0
            up = 0
    
            for index, dxy in enumerate(dict_dxy):
                if check_dxy_cond(dxy_case, dxy_threshold, dxy["비율"]):
                    if index < dict_dxy_max_index: # 예외처리 : 인덱스가 리스트 넘어가는 것 방지
                        try: # 예외처리 : 휴일 나스닥 데이터 접근 제외
                            next_index = index + 1
                            next_date = dict_dxy[next_index]["날짜"]
                            next_ixic = df_ixic.loc[next_date] # 예외발생 가능
                
                            total = total + 1
                            if next_ixic["비율"] >= ixic_threshold:
                                up = up + 1
                        except:
                            pass # no data

            if total != 0:
                up_per = round((up/total), 2) * 100

                if up_per > 0:
                    print(f"결과: 달러 {dxy_threshold}% 이상 {dxy_case} 시, 다음날 나스닥 {ixic_threshold}% 이상 상승 확률 :", up_per, f"({up}/{total})")
                    is_printed = True
            else:
                pass # no data
        if is_printed:
            print()

결과: 달러 0% 이상 상승 시, 다음날 나스닥 0% 이상 상승 확률 : 53.0 (69/130)
결과: 달러 0% 이상 상승 시, 다음날 나스닥 1% 이상 상승 확률 : 21.0 (27/130)
결과: 달러 0% 이상 상승 시, 다음날 나스닥 2% 이상 상승 확률 : 4.0 (5/130)

결과: 달러 1% 이상 상승 시, 다음날 나스닥 0% 이상 상승 확률 : 67.0 (2/3)
결과: 달러 1% 이상 상승 시, 다음날 나스닥 1% 이상 상승 확률 : 33.0 (1/3)
결과: 달러 1% 이상 상승 시, 다음날 나스닥 2% 이상 상승 확률 : 33.0 (1/3)

결과: 달러 0% 이상 하락 시, 다음날 나스닥 0% 이상 상승 확률 : 59.0 (73/123)
결과: 달러 0% 이상 하락 시, 다음날 나스닥 1% 이상 상승 확률 : 22.0 (27/123)
결과: 달러 0% 이상 하락 시, 다음날 나스닥 2% 이상 상승 확률 : 6.0 (7/123)
결과: 달러 0% 이상 하락 시, 다음날 나스닥 3% 이상 상승 확률 : 1.0 (1/123)

결과: 달러 1% 이상 하락 시, 다음날 나스닥 0% 이상 상승 확률 : 100.0 (4/4)
결과: 달러 1% 이상 하락 시, 다음날 나스닥 1% 이상 상승 확률 : 25.0 (1/4)



- 실제 사용할만한 의미 있는 결과 (확률이 높거나, 수익률이 높거나) : <b>달러 1% 이상 하락 시, 다음날 나스닥 0% 이상 상승 확률 : 100.0 (4/4)</b>
  - 하지만 기회가 몇 번 없음
  - 하지만 수익률이 낮음 -> 나스닥 100? TQQQ? 코스닥?
  - 하지만 실제로 가능할까? (혹시 장 시작부터 꼭대기에서 시작해서 못 사는거 아닐까?)
- 향후 진행 방향
  - 다른 조건(패턴) 추가 : 연속성 (<- 나스닥 많이 오른날 기준으로 분석 진행)

In [13]:
df_ixic[(df_ixic["상승"])&(df_ixic["비율"]>=2)]

날짜         종가         시가         고가         저가  \
날짜                                                                       
2023- 01- 06  2023- 01- 06  10,569.29  10,363.96  10,604.14  10,265.04   
2023- 01- 20  2023- 01- 20  11,140.43  10,924.66  11,143.17  10,885.64   
2023- 01- 23  2023- 01- 23  11,364.41  11,171.94  11,405.50  11,144.03   
2023- 02- 01  2023- 02- 01  11,816.32  11,573.14  11,904.02  11,500.33   
2023- 02- 02  2023- 02- 02  12,200.82  12,065.15  12,269.55  12,024.14   
2023- 03- 14  2023- 03- 14  11,428.15  11,357.74  11,467.01  11,284.93   
2023- 03- 16  2023- 03- 16  11,717.28  11,384.87  11,733.80  11,365.49   
2023- 04- 27  2023- 04- 27  12,142.24  11,972.15  12,154.01  11,950.92   
2023- 05- 05  2023- 05- 05  12,235.41  12,073.03  12,264.83  12,065.72   
2023- 05- 26  2023- 05- 26  12,975.69  12,736.42  13,001.91  12,729.74   
2023- 11- 10  2023- 11- 10  13,798.11  13,571.19  13,802.53  13,556.35   
2023- 11- 14  2023- 11- 14  14,094.38  14,015.37  14,124.06  14,003.14   

                  거래량   변동 %    상승     하락    비율  
날짜                                               
2023- 01- 06  970.67M  2.56%  True  False  2.56  
2023- 01- 20    1.16B  2.66%  True  False  2.66  
2023- 01- 23    1.05B  2.01%  True  False  2.01  
2023- 02- 01    1.21B  2.00%  True  False  2.00  
2023- 02- 02    1.33B  3.25%  True  False  3.25  
2023- 03- 14    1.18B  2.14%  True  False  2.14  
2023- 03- 16    1.13B  2.48%  True  False  2.48  
2023- 04- 27  986.43M  2.43%  True  False  2.43  
2023- 05- 05    1.01B  2.25%  True  False  2.25  
2023- 05- 26  914.77M  2.19%  True  False  2.19  
2023- 11- 10  998.71M  2.05%  True  False  2.05  
2023- 11- 14    1.11B  2.37%  True  False  2.37

- 분석 결과
    - 2023-01-06 +2.56%
    	- 달러 상승 + 나스닥 크게 하락(-1.47%)
    - 2023-01-20 +2.66%
    	- 달러 2일 연속 하락 + 나스닥 크게 하락 + 2일 연속 하락(-1.24%, -0.96%)
    - 2023-01-23 +2.01%
    	- 달러 3일 연속 하락 + 나스닥 크게 상승(+2.66%)
    - 2023-02-01 +2.00%
    	- 달러 하락 + 나스닥 크게 상승(+1.67%)
    - 2023-02-02 +3.25%
    	- 달러 2일 연속 하락 + 나스닥 크게 상승(+2.00%)
    - 2023-03-14 +2.14%
    	- 달러 3일 연속 하락 + 나스닥 상승
    - 2023-03-16 +2.48%
    	- 달러 2일 연속 상승 + 나스닥 연속 3일 상승
    - 2023-04-27 +2.43%
    	- 달러 하락 + 나스닥 상승
    - 2023-05-05 +2.25%
    	- 달러 상승 + 나스닥 4일 연속 하락
    - 2023-05-26 +2.19%
        - 달러 3일 연속 상승 + 나스닥 크게 상승(+1.71%)
    - 2023-11-10 +2.05%
        - 달러 4일 연속 상승 + 나스닥 하락
    - 2023-11-14 +2.37%
        - 달러 2일 연속 하락 + 나스닥 하락
- 결론
    - 나스닥 조건 추가
    - 달러/나스닥 연속성 조건 추가

In [22]:
# 분석 결과를 참고하여 "분석 1" 개선 -> "분석 2"
# (추가) 거래량 조건 추가, 다른 통계량 추가(평균 ...), 다른 데이터 단위 추가 후 앙상블, 전체 과거 데이터 적용
# (점점 창우가 하고자 하는 것과 비슷해지는 모습)